[healthcheckextensionv2] Introduce health check extension based on component status reporting

mwear opened the pull request:
**Description:**
This PR introduces a candidate to replace the current health check extension. This extension is based on component status reporting and provides insight into collector health at the pipeline and component level. It includes an HTTP service and a gRPC service. The HTTP service has a status endpoint that provides the ability to probe the overall collector health and the health of individual pipelines. It also provides a config endpoint to obtain the config of the currently running collector.

The gRPC service is an implementation of the [grpc_health_v1 service](https://github.com/grpc/grpc-proto/blob/master/grpc/health/v1/health.proto). While not as detailed as the HTTP service, it provides the ability to check the health of the collector overall and the health of individual pipelines via a unary or streaming RPC.

Looking at the README in this PR is a good place to start to understand the changes. 

I have temporarily named this extension healthcheckv2. We can discuss the best way to integrate this work going forward, but my reasoning for using a different name is to facilitate the code review (the diff would be hard to follow if we were to overwrite the current extension). If we decide to move forward with this extension and replace the existing one, we can handle the deletion of the old extension and the renaming in a followup PR.

Ultimately this extension will need some form of one of the following before it will be a suitable replacement:
- https://github.com/open-telemetry/opentelemetry-collector/pull/8684
- https://github.com/open-telemetry/opentelemetry-collector/pull/8788 

The more components that adopt status reporting, the more useful this extension will be.

**Link to tracking Issue:** #26661

**Testing:** Units, Manual

**Documentation:** README

mwear committed:
Introduce health check extension based on component status reporting

codeboten commented:
Thanks for all this work @mwear! 

> Ultimately this extension will need some form of one of the following before it will be a suitable replacement:

Can you elaborate on why that's needed? My understanding of the current healthcheck extension is that it doesn't actually work for checking the health of pipelines (see issue: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/11780 and the readme warning: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/healthcheckextension#health-check). Is the implementation in this PR not already suitable as a replacement in that case?

I'm asking to better understanding if the PRs listed are necessary before moving this work forward.

mwear commented:
As you point out, there are currently some issues with the accuracy of the current health check extension. Beyond this, we now have component status reporting as a collector feature, which allows us to report status of individual components. The overall collector health and pipeline health can be derived by looking at the health of the underlying components. All components have a basic level of automated status reporting, but there are nuanced failure and recovery cases that need to be handled on a case by case basis. That is, some components will need to do some level of manual status reporting for their health to be observable. The two PRs I mentioned add manual status reporting to the core exporters based on error conditions they can encounter. They attempt to identify the RecoverableError conditions, when and if they have recovered, and the PermanentError conditions that will require human intervention to fix. Without this information, the health check extension will be able to tell you very little about the health of your exporters. You will be able to know if they are running, or not, at that's all. tl;dr: we need to add status reporting to components that we care about for purposes of collector health. Exporters are the highest priority, then receivers and processors

codeboten commented on .github/CODEOWNERS:
```diff
@@ -93,6 +93,7 @@ extension/encoding/textencodingextension/                @open-telemetry/collect
 extension/encoding/zipkinencodingextension/              @open-telemetry/collector-contrib-approvers @MovieStoreGuy @dao-jun
 extension/headerssetterextension/                        @open-telemetry/collector-contrib-approvers @jpkrohling
 extension/healthcheckextension/                          @open-telemetry/collector-contrib-approvers @jpkrohling
+extension/healthcheckextensionv2/                        @open-telemetry/collector-contrib-approvers @mwear
```
I'd much rather this replaces the existing extension, than to add a new extension :) 

But i appreciate your comment about making it more difficult to review that way

codeboten commented:
> You will be able to know if they are running, or not, at that's all.

Fair enough, though I would say at this stage, it's more than the current healthcheck extension's capabilities

mwear commented:
> > You will be able to know if they are running, or not, at that's all.
> 
> Fair enough, though I would say at this stage, it's more than the current healthcheck extension's capabilities

I agree. I suppose we can classify those PRs as nice to have in the short term and must have in the longer term.

vynu commented:
@mwear i've tried your healhcheck-v2 code to build my own otel collector (v0.92.0) build failing with following errors:

`280.3 Error: failed to compile the OpenTelemetry Collector distribution: go subcommand failed with args '[build -trimpath -o otelcol -ldflags=-s -w]': exit status 1. Output:
280.3 # go.opentelemetry.io/collector/exporter/exporterhelper
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/exporter@v0.92.0/exporterhelper/obsexporter.go:65:55: undefined: obsmetrics.TagKeyExporter
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/exporter@v0.92.0/exporterhelper/obsexporter.go:231:28: undefined: obsmetrics.ExporterSentSpans
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/exporter@v0.92.0/exporterhelper/obsexporter.go:232:30: undefined: obsmetrics.ExporterFailedToSendSpans
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/exporter@v0.92.0/exporterhelper/obsexporter.go:234:28: undefined: obsmetrics.ExporterSentMetricPoints
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/exporter@v0.92.0/exporterhelper/obsexporter.go:235:30: undefined: obsmetrics.ExporterFailedToSendMetricPoints
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/exporter@v0.92.0/exporterhelper/obsexporter.go:237:28: undefined: obsmetrics.ExporterSentLogRecords
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/exporter@v0.92.0/exporterhelper/obsexporter.go:238:30: undefined: obsmetrics.ExporterFailedToSendLogRecords
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/exporter@v0.92.0/exporterhelper/obsexporter.go:289:30: undefined: obsmetrics.ExporterFailedToEnqueueSpans
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/exporter@v0.92.0/exporterhelper/obsexporter.go:291:30: undefined: obsmetrics.ExporterFailedToEnqueueMetricPoints
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/exporter@v0.92.0/exporterhelper/obsexporter.go:293:30: undefined: obsmetrics.ExporterFailedToEnqueueLogRecords
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/exporter@v0.92.0/exporterhelper/obsexporter.go:293:30: too many errors
280.3 # go.opentelemetry.io/collector/processor/processorhelper
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/processor@v0.92.0/processorhelper/obsreport.go:77:58: undefined: obsmetrics.TagKeyProcessor
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/processor@v0.92.0/processorhelper/obsreport.go:192:32: undefined: obsmetrics.ProcessorAcceptedSpans
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/processor@v0.92.0/processorhelper/obsreport.go:193:31: undefined: obsmetrics.ProcessorRefusedSpans
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/processor@v0.92.0/processorhelper/obsreport.go:194:31: undefined: obsmetrics.ProcessorDroppedSpans
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/processor@v0.92.0/processorhelper/obsreport.go:196:32: undefined: obsmetrics.ProcessorAcceptedMetricPoints
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/processor@v0.92.0/processorhelper/obsreport.go:197:31: undefined: obsmetrics.ProcessorRefusedMetricPoints
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/processor@v0.92.0/processorhelper/obsreport.go:198:31: undefined: obsmetrics.ProcessorDroppedMetricPoints
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/processor@v0.92.0/processorhelper/obsreport.go:200:32: undefined: obsmetrics.ProcessorAcceptedLogRecords
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/processor@v0.92.0/processorhelper/obsreport.go:201:31: undefined: obsmetrics.ProcessorRefusedLogRecords
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/processor@v0.92.0/processorhelper/obsreport.go:202:31: undefined: obsmetrics.ProcessorDroppedLogRecords
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/processor@v0.92.0/processorhelper/obsreport.go:202:31: too many errors
280.3 # go.opentelemetry.io/collector/receiver/receiverhelper
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/receiver@v0.92.0/receiverhelper/obsreport.go:76:26: undefined: obsmetrics.TagKeyReceiver
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/receiver@v0.92.0/receiverhelper/obsreport.go:77:26: undefined: obsmetrics.TagKeyTransport
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/receiver@v0.92.0/receiverhelper/obsreport.go:306:32: undefined: obsmetrics.ReceiverAcceptedSpans
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/receiver@v0.92.0/receiverhelper/obsreport.go:307:31: undefined: obsmetrics.ReceiverRefusedSpans
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/receiver@v0.92.0/receiverhelper/obsreport.go:309:32: undefined: obsmetrics.ReceiverAcceptedMetricPoints
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/receiver@v0.92.0/receiverhelper/obsreport.go:310:31: undefined: obsmetrics.ReceiverRefusedMetricPoints
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/receiver@v0.92.0/receiverhelper/obsreport.go:312:32: undefined: obsmetrics.ReceiverAcceptedLogRecords
280.3 /app/go/pkg/mod/go.opentelemetry.io/collector/receiver@v0.92.0/receiverhelper/obsreport.go:313:31: undefined: obsmetrics.ReceiverRefusedLogRecords
280.3 
------
failed to solve: process "/bin/sh -c builder --config builder-config.yaml" did not complete successfully: exit code: 1
`

here is my build-config.yaml
`dist:
  name: otelcol
  description: Custom OTel Collector distribution
  output_path: ./otelcol-dev
  otelcol_version: 0.92.0

extensions:
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/extension/healthcheckextensionv2 v0.92.0

exporters:
  - gomod: go.opentelemetry.io/collector/exporter/otlphttpexporter v0.92.0

  - gomod: go.opentelemetry.io/collector/exporter/otlpexporter v0.92.0

  - gomod: go.opentelemetry.io/collector/exporter/debugexporter v0.92.0

  - gomod: go.opentelemetry.io/collector/exporter/loggingexporter v0.92.0

  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/exporter/fileexporter v0.92.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/exporter/splunkhecexporter v0.92.0

processors:
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/processor/attributesprocessor v0.92.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/processor/resourcedetectionprocessor v0.92.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/processor/filterprocessor v0.92.0
  - gomod: go.opentelemetry.io/collector/processor/batchprocessor v0.92.0

receivers:
  - gomod: go.opentelemetry.io/collector/receiver/otlpreceiver v0.92.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/filelogreceiver v0.92.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusreceiver v0.92.0

replaces:
  - github.com/open-telemetry/opentelemetry-collector-contrib/extension/healthcheckextensionv2 => ../healthcheckextensionv2

`
i am using below builder version
`go.opentelemetry.io/collector/cmd/builder@v0.92.0`

help..

codeboten commented:
@vynu just updated your comment to make the error messages more readable :)

vynu commented:
thanks @codeboten i've been trying that .. but for some reason i am seeing error while editing ..

mwear committed:
Refactor aggregator

There were previously different types for CollectorStatus and PipelineStatus
that have been unified with a newly introduced AggregateStatus type. It cleaned
up a lot of special cases.

mwear committed:
Make gotidy

vynu commented:
thanks for the suggestion @mwear  i've successful build and tested on my laptop and also in AWS fargate .. worked just fine ..

one interesting error i see in cloudwatch logs:

`loopyWriter exiting with error: transport closed by client\t{\"grpc_log\": true}`

this happening  whenever ALB performs health check calls on tasks ..  i know ALB target group Grpc health check are little weird .. any suggestion on this ?

jpkrohling commented:
@mwear, is this ready for review? There was some discussions during yesterday's SIG call but I'm not sure if the state of the PR reflects what was discussed yesterday. If it does, let me know and I'll review this PR.

mwear commented:
My takeaway from the SIG meeting is that folks would like pipeline health monitoring to be optional. I plan to update this PR to make it optional, but right now, it's the only option. If you want to review this PR with that in mind, feel free. I will let everyone know when the additional functionality is implemented.

codeboten commented on extension/healthcheckextensionv2/config.go:
```diff
@@ -0,0 +1,36 @@
+	}
+
+	if c.HTTPSettings != nil && c.HTTPSettings.Endpoint == "" {
+		return errors.New("healthcheck extension: grpc endpoint required")
```
```suggestion
		return errors.New("healthcheck extension: http endpoint required")
```

codeboten commented on extension/healthcheckextensionv2/config_test.go:
```diff
@@ -0,0 +1,91 @@
+		},
+	} {
+		err := tc.config.Validate()
+		assert.Equal(t, tc.valid, err == nil)
```
This could be an assertion of the error instead

mwear commented:
> thanks for the suggestion @mwear i've successful build and tested on my laptop and also in AWS fargate .. worked just fine ..
> 
> one interesting error i see in cloudwatch logs:
> 
> `loopyWriter exiting with error: transport closed by client\t{\"grpc_log\": true}`
> 
> this happening whenever ALB performs health check calls on tasks .. i know ALB target group Grpc health check are little weird .. any suggestion on this ?

@vynu, I can look into this. Do you happen to know if ALB uses the `Check` or `Watch` RPC?

mwear committed:
Update extension/healthcheckextensionv2/config.go

Co-authored-by: Alex Boten <alex@boten.ca>

codeboten commented on extension/healthcheckextensionv2/internal/testhelpers/helpers.go:
```diff
@@ -0,0 +1,75 @@
+	}
+}
+
+// Returns a map of hypothetical pipelines identified by their stringified typeVal.
```
```suggestion
// NewPipelines returns a map of hypothetical pipelines identified by their stringified typeVal.
```

codeboten commented on extension/healthcheckextensionv2/internal/testhelpers/helpers.go:
```diff
@@ -0,0 +1,75 @@
+	return []*component.InstanceID{p.ReceiverID, p.ProcessorID, p.ExporterID}
+}
+
+// Returns a metadata for a hypothetical pipeline.
```
```suggestion
// NewPipelineMetadata returns a metadata for a hypothetical pipeline.
```

codeboten commented on extension/healthcheckextensionv2/internal/http/serialization.go:
```diff
@@ -0,0 +1,105 @@
+
+func isHealthy(ev *component.StatusEvent, now time.Time, recoveryDuration time.Duration) bool {
+	if ev.Status() == component.StatusRecoverableError &&
+		now.Compare(ev.Timestamp().Add(recoveryDuration)) == -1 {
```
Could this be replaced with `.Before`?

codeboten commented on extension/healthcheckextensionv2/internal/http/serialization.go:
```diff
@@ -0,0 +1,105 @@
+	Timestamp    time.Time `json:"status_time"`
+}
+
+var stringToStatusMap = map[string]component.Status{
```
is this map needed? the reverse mapping is already defined in the `String` method, wouldn't want the two to diverge

https://github.com/open-telemetry/opentelemetry-collector/blob/d2e35bdeb7ec040ca5a81e890a754b362eb99a02/component/status.go#L25

mwear replied:
The reason this is here is that it is very convenient if a struct has the property that if you `json.Marshal` it, then `json.Unmarshal` it you get the same value. Originally, I had an unexported `component.Status` on `serializeableStatus` so that it was not present in the JSON, however, when unmarshaling, you will always get the zero value: `component.StatusNone` and not the original status. Defining `Status` as method and using this map fixes that problem. I'd prefer to keep it. We could consider adding to core if we're worried about divergence.

codeboten commented on extension/healthcheckextensionv2/internal/status/aggregator.go:
```diff
@@ -0,0 +1,227 @@
+)
+
+const (
+	ScopeAll        = ""
```
why `""` and not `"all"`?

mwear replied:
This is aligned with how the gRPC healthcheck works: https://github.com/grpc/grpc-go/blob/master/health/grpc_health_v1/health_grpc.pb.go#L47-L50. Also, more practically, I'd end up adding a lot of conditions checking for empty string, and replacing it with `all`. eg.

```go
        pipeline := r.URL.Query().Get("pipeline")
	st, ok := s.aggregator.AggregateStatus(pipeline, s.settings.Status.Detailed)
	if !ok {
		w.WriteHeader(http.StatusNotFound)
		return
	}
```

Would have to become:
```go
        pipeline := r.URL.Query().Get("pipeline")
        if pipeline == "" {
            pipeline = "all"
        }
	st, ok := s.aggregator.AggregateStatus(pipeline, s.settings.Status.Detailed)
	if !ok {
		w.WriteHeader(http.StatusNotFound)
		return
	}
```

This would be the case in each of the http handlers and RPCs.

mwear replied:
I added `Scope` and `Detail` types to the aggregator. It retains the usability of the string / bool types, but makes them easier to understand and less error prone. Here are some examples of how it looks now:

```go
pipeline := r.URL.Query().Get("pipeline")
st, ok := s.aggregator.AggregateStatus(
    status.Scope(pipeline),
    status.Detail(s.settings.Status.Detailed),
)
```

```go
st, ok := agg.AggregateStatus(status.ScopeAll, status.IncludeSubtrees)
```

codeboten commented on extension/healthcheckextensionv2/internal/status/aggregator.go:
```diff
@@ -0,0 +1,227 @@
+	return result
+}
+
+func kindToString(k component.Kind) string {
```
Kind's `String` method does the same thing

mwear commented on extension/healthcheckextensionv2/README.md:
````diff
@@ -0,0 +1,337 @@
+If the same request is made to a collector with `http.status.detailed` to `false`, you will only get
+the overall status. The pipeline and component level statuses will be omitted.
+
+```
````
```suggestion
```json
```

mwear committed:
Apply suggestions from code review

Co-authored-by: Alex Boten <alex@boten.ca>

mwear committed:
Implement suggestions from review

mwear commented on extension/healthcheckextensionv2/metadata.yaml:
```diff
@@ -0,0 +1,9 @@
+  class: extension
+  stability:
+    development: [extension]
+  distributions: [contrib]
```
```suggestion
  distributions: [core, contrib]
```

mwear commented on extension/healthcheckextensionv2/internal/grpc/config.go:
```diff
@@ -0,0 +1,11 @@
+
+type Settings struct {
+	configgrpc.GRPCServerSettings `mapstructure:",squash"`
+	Debug                         bool `mapstructure:"debug"`
```
Server reflection allows me to easily use grpcurl to test the grpc service. I can probably remove this and just use the proto source file for testing. If it's here to stay, I'll find a better place for it (like on GRPCServerSettings).

mwear committed:
Add Scope and Detail types to Aggregator

TylerHelmuth commented on extension/healthcheckextensionv2/README.md:
I'd love to see how these features/improvements could be incorporated into the existing health check extension instead of introducing a new component.

mwear replied:
Part of the reason this was done separately is because it's a pretty big overhaul of the existing extension. Based on the SIG discussion, the plan will be to at least fuse them so there are not separate extensions. We may want to deprecate some functionality from the existing extension, but there is some we would like carry forward as well. Given the size of this PR and additional scope creep I am working through, it might make sense to temporarily introduce this as a separate component and handle the fusing separately.

TylerHelmuth replied:
I am definitely in favor of a fuse-and-then-replace strategy. The ultimate result would be this component, but we'd get to reuse the existing `health_check` name, saving lots of docs updates and hopefully saving end users from breaking config changes (since most users only using `health_check:` and not doing more configuration OR are using the helm chart which can handle things gracefully)

I think one way to start fusing this into the existing extension is to leave this PR open as reference, and then start introducing PRs to health check extension that add the config/functionality.  I think we should stick with our new-component-model of submitting separate PRs for config and business logic; that should help make the PRs smaller.

mwear replied:
Once this PR has all the functionality desired, I'll work with folks to come up with an integration plan. I think under the hood, there will be two extensions, and depending on config, we will wire up one or the other, and over time try to pare that down to a single extension.

mwear committed:
Use pipeline status for availability during start/shutdown

mwear committed:
Multiple health strategies for http

mwear committed:
Multiple health strategies via different aggregation funcs

pratik2294 commented:
Hey all, I really need to use the health check extension for K8s environment but can't use it due to #11780 & I'm wondering when will be the health_checkv2 be available with otel collector.

mwear commented:
@pratik2294, I am actively working on this. Below is a summary of what needs to be done. The first point is pretty close to being ready. 

- Make opting in to component health optional
  - This is analogous to the `check_collector_pipeline`
  - There will be possible to opt-in to recoverable errors only, permanent errors only, or both
- Rather than introduce this as v2 there is a desire to replace the current extension with minimal friction, this is doable, but the details need to be decided on.
- In order to monitor exporter failures some version of the following will need to merge:
  - https://github.com/open-telemetry/opentelemetry-collector/pull/8684
  - https://github.com/open-telemetry/opentelemetry-collector/pull/8788

mwear committed:
Simplify HTTP strategies

mwear committed:
gRPC support for component health options

mwear committed:
Fusion

Support config and responses from current healthcheck extension.

mwear committed:
Update deps / rename settings -> config

mwear committed:
Readme fusion

mwear commented:
This PR has been updated and the latest round of improvements includes:
- Support for v1 config and responses
  - This extension should be a seamless replacement for the existing one
  - We will need to figure out merge / rename or other strategy for integration
  - The `check_collector_pipeline` config remains for backwards compatibility, but there is no functionality backing it. Users are encouraged to use v2 for similar functionality.
- By default error status events are not factored into the health check
  - An aggregate status is always available in the response body, but not reflected in the response code unless explicitly enabled
  - Users can opt into recoverable errors, permanent errors, or both.

jpkrohling reviewed (approved):
I reviewed everything, except the aggregation logic. Apart from a few requests around better observability, I'm willing to give this a try and get real-world feedback.

mwear committed:
Update extension/healthcheckextensionv2/README.md

Co-authored-by: Juraci Paixão Kröhling <juraci.github@kroehling.de>

mwear commented on extension/healthcheckextensionv2/extension.go:
```diff
@@ -0,0 +1,211 @@
+
+// Start implements the component.Component interface.
+func (hc *healthCheckExtension) Start(ctx context.Context, host component.Host) error {
+	hc.telemetry.Logger.Info("Starting health check extension V2", zap.Any("config", hc.config))
```
I can make this change, but it currently matches the behavior of the existing health check extension: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/0b07906f4b50e70430fb868db8d464104a92f6c1/extension/healthcheckextension/healthcheckextension.go#L35

mwear commented on extension/healthcheckextensionv2/README.md:
````diff
@@ -0,0 +1,414 @@
+```yaml
+extensions:
+  healthcheckv2:
+    use_v2: true
````
The intention of this flag is for it to be temporary and ease the transition to v2. I plan to remove it after we can remove the v1 functionality. It's already setup for this as the flag is implemented on the `LegacyConfig` struct, all of which is slated for future removal. If you would still like this change, let me know.

mwear committed:
Address code review feedback

- Words of caution on the readme
- Remove prefix on errors
- Log details on discarded events

mwear committed:
Address more feedback; additional cleanup

mwear commented on extension/healthcheckextensionv2/extension.go:
```diff
@@ -0,0 +1,211 @@
+		case <-hc.eventCh:
+			hc.telemetry.Logger.Info("healthcheck: discarding event received after shutdown")
+		case <-ctx.Done():
+			return
```
`StatusEvents` come in via calls to the `StatusWatcher` optional interface and are written to an unbuffered channel. The interface can be invoked by the collector before the extension starts and after it has stopped. We need to ensure to read all events written to the channel so future writes do not block. If we were to time this context out early, subsequent calls to the `StatusWatcher` interface would block indefinitely, which would be a problem.

mwear commented on extension/healthcheckextensionv2/internal/grpc/grpc.go:
```diff
@@ -0,0 +1,130 @@
+				},
+			)
+			if err != nil {
+				return grpcstatus.Error(codes.Canceled, "Stream has ended.")
```
I believe that `stream.Send` can fail in situations where we would not receive a cancellation via `context.Done`. This follows the same pattern as the grpc health server implementation:
https://github.com/grpc/grpc-go/blob/815e2e2d358ff8d6f5e67098271ccec1a2546961/health/server.go#L99-L102

jpkrohling replied:
Right, that's why I think the messages need to be different. When someone is reporting an error to you and the message says "Stream has ended", I think it would be useful to differentiate between this case here and the other place where the same message is being used.

mwear commented on extension/healthcheckextensionv2/README.md:
```diff
@@ -0,0 +1,414 @@
+
+#### Status Endpoint
+
+The HTTP service provides a status endpoint that can be probed for overall collector status and
```
I added a cautionary statement here and for the config endpoint. Let me know if it needs any more wordsmithing.

mwear commented on extension/healthcheckextensionv2/extension.go:
```diff
@@ -0,0 +1,211 @@
+	return err
+}
+
+func (hc *healthCheckExtension) eventLoop(ctx context.Context) {
```
👍 on the observability suggestions. Would it be OK to address them in a followup PR as this one is already quite large?

jpkrohling replied:
Absolutely

jpkrohling commented:
This PR LGTM and can be merged as is, nothing I commented before is blocking and the accepted suggestions can be made in follow-up PRs.

mwear committed:
Post rebase fixes

valageri commented:
@mwear When can this PR be merged?

gjshao44 commented:
Thanks @mwear for the extensive work to get us here. Now all checks have passed and code approved, could the branch be merged? @SurenNihalani  @codeboten @yurishkuro

atoulme commented on extension/healthcheckextensionv2/metadata.yaml:
```diff
@@ -0,0 +1,9 @@
+  class: extension
+  stability:
+    development: [extension]
+  distributions: [core, contrib]
```
this is not true yet, right?

atoulme commented on extension/healthcheckextensionv2/metadata.yaml:
```diff
@@ -0,0 +1,9 @@
+    development: [extension]
+  distributions: [core, contrib]
+  codeowners:
+    active: [mwear]
```
missing jpkrohling

atoulme commented on .github/ISSUE_TEMPLATE/bug_report.yaml:
```diff
@@ -93,6 +93,7 @@ body:
       - extension/encoding/zipkinencoding
       - extension/headerssetter
       - extension/healthcheck
+      - extension/healthcheckextensionv2
```
Sounds like it should be:
```suggestion
      - extension/healthcheckv2
```

atoulme commented on extension/healthcheckextensionv2/config.go:
```diff
@@ -0,0 +1,99 @@
+		if c.HTTPConfig.Endpoint == "" {
+			return errHTTPEndpointRequired
+		}
+		if c.HTTPConfig.Status.Enabled && !strings.HasPrefix(c.HTTPConfig.Status.Path, "/") {
```
Is empty path not valid?

mwear replied:
An explicitly empty path is not valid. If it is unset, you will get the default value supplied via the factory. This is the same behavior as the existing extension: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/6be6423d34121953c2edd1577d6d9f5a9d44585b/extension/healthcheckextension/config.go#L61-L63.

atoulme commented on extension/healthcheckextensionv2/config.go:
```diff
@@ -0,0 +1,99 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package healthcheckextensionv2 // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/healthcheckextensionv2"
```
```suggestion
package healthcheckv2extension // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/healthcheckv2extension"
```
I believe the name should be suffixed with extension

mwear committed:
Post rebase fixes

mwear committed:
Rename healthcheckextensionv2 -> healthcheckv2extension

mwear committed:
Update metadata

mwear commented:
We discussed next steps for this PR at the collector SIG on Wednesday. We know we want to replace the existing extension with this updated version, but there was a question as to whether we do that directly, or do it in steps. We decided we will do this in steps. We'll integrate this as healthcheckv2 and when we're confident in it, we will replace the existing extension and delete healthcheckv2. healthcheckv2 will be in development stability for its lifetime and interested parties can use the collector builder to test it out. The next point of discussion was the size of this PR. This needs more approvals before it can merge. If nobody is willing to review it in its current state, then I will have to split this up into smaller pieces to facilitate the review process. I'd like to know if there is any interest in reviewing this as-is before heading down that path.

nathanhammond commented:
Hey, so I've attempted to test this in production via https://github.com/scorecard-ai/scorecard-otel-collector, here are a few things I noticed.

1. It appears that `status_time` is ending up as a fixed value. I originally expected to see uptime from the original `health_check`.

```
[ec2-user@ip-172-31-NN-NNN ~]$ curl http://0.0.0.0:13133
{"start_time":"2024-04-03T11:42:45.211686482Z","healthy":true,"status":"StatusOK","status_time":"2024-04-03T11:42:45.572109156Z"}
[ec2-user@ip-172-31-NN-NNN ~]$ curl http://0.0.0.0:13133
{"start_time":"2024-04-03T11:42:45.211686482Z","healthy":true,"status":"StatusOK","status_time":"2024-04-03T11:42:45.572109156Z"}
[ec2-user@ip-172-31-NN-NNN ~]$ curl http://0.0.0.0:13133
{"start_time":"2024-04-03T11:42:45.211686482Z","healthy":true,"status":"StatusOK","status_time":"2024-04-03T11:42:45.572109156Z"}
[ec2-user@ip-172-31-NN-NNN ~]$ curl http://0.0.0.0:13133
{"start_time":"2024-04-03T11:42:45.211686482Z","healthy":true,"status":"StatusOK","status_time":"2024-04-03T11:42:45.572109156Z"}
[ec2-user@ip-172-31-NN-NNN ~]$ curl http://0.0.0.0:13133
```

2. Not specifying `path: "/"` for `http` results in the server sending 404s for root, not the status.

3. GRPC calls are failing:
```
[ec2-user@ip-172-31-NN-NNN ~]$ grpcurl -vv -plaintext 0.0.0.0:13132 grpc.health.v1.Health/Check 
Error invoking method "grpc.health.v1.Health/Check": failed to query for service descriptor "grpc.health.v1.Health": server does not support the reflection API
```

My config is this:
```
extensions:
  healthcheckv2:
    use_v2: true
    grpc:
      endpoint: "0.0.0.0:13132"
      transport: "tcp"
    http:
      endpoint: "0.0.0.0:13133"
      status:
        enabled: true
        path: "/"
      config:
        enabled: false
```

mwear commented:
@nathanhammond, thank you so much for trying this out and providing feedback. I have answers for most of your questions below.

> It appears that status_time is ending up as a fixed value. I originally expected to see uptime from the original health_check.

`status_time` is the time of the underlying status event generated by calls to the component status API. In general, this is not a fixed time, but right now status is only being reported during startup and shutdown. The value you are seeing is the time of the `StatusOK` event during startup. In the future, components will report runtime status and the timestamp will update on changes in status. Uptime can be computed by subtracting `start_time` from now.

> GRPC calls are failing, but this could be my fault where I don't know what I'm supposed to be doing. (No real documentation for configuring this.)

Reflection is not enabled on the GRPC server, so you will need to point `grpcurl` to the protos. You can clone: https://github.com/grpc/grpc then use 

```
grpcurl -import-path grpc/src/proto/grpc/health/v1 -proto health.proto -plaintext -d '{"service":""}' localhost:13132 grpc.health.v1.Health.Check
or
grpcurl -import-path grpc/src/proto/grpc/health/v1 -proto health.proto -plaintext -d '{"service":""}' localhost:13132 grpc.health.v1.Health.Watch
```

> http.status.detailed: false throws a marshaling error.

This was a bug on the readme. I've fixed it. Thanks for pointing it out. For this behavior you can pass `verbose` as a query parameter to the http service. For example: http://0.0.0.0:13133/status?verbose.

nathanhammond commented:
In my current use case I am trying to enable health checks from AWS's Application Load Balancer and (in spite of it being configured "correctly" to the best of my knowledge) I can't get either of the HTTP or gRPC health checks to respond in a way that the load balancer is happy keeping the service in rotation.

I'm going to put more tracing in place to figure out why it's not working, but I suspect the gRPC request from the ALB is not going to match Google's pattern, so we'll see how that goes.

(that might become a feature request, but I can always add it to my fork)

mwear commented:
I'd be interested to know what you find and definitely happy to take feature requests to address in follow up PRs. This PR has been a victim of scope creep and is likely going to have to be broken into smaller PRs to facilitate reviews.

mwear commented:
@vynu were you able to get this working in AWS when you tested it? If so, do you have any suggestions for @nathanhammond?

nathanhammond commented:
The update here is I have HTTP working, but gRPC doesn't respond to the AWS ALB request. I'll prep that in my fork so we can at least see what it looks like.

nathanhammond commented:
Last update: I now have this extension working behind an AWS ALB and correctly responding to both HTTP and gRPC healthchecks. Review: 10/10. Will be shipped to (low-traffic) production on Monday. (Already "in prod", just not in use.)

***

```js
import * as elb from 'aws-cdk-lib/aws-elasticloadbalancingv2';

// GRPC requests.
const grpcListener = loadBalancer.addListener('otlp-grpc', {
  port: 4317,
  protocol: elb.ApplicationProtocol.HTTPS,
  open: true
});
grpcListener.addCertificates(`otlp-grpc-certificate`, [certificate]);

grpcListener.addTargets('otlp-grpc-target', {
  port: 4317,
  protocol: elb.ApplicationProtocol.HTTP,
  protocolVersion: elb.ApplicationProtocolVersion.GRPC,
  targets: [autoScalingGroup],
  healthCheck: {
    protocol: elb.Protocol.HTTP,
    port: '13132',
    path: '/grpc.health.v1.Health/Check',
    healthyGrpcCodes: '0'
  }
});

// HTTP requests.
const httpsListener = loadBalancer.addListener('otlp-http', {
  port: 4318,
  protocol: elb.ApplicationProtocol.HTTPS,
  open: true
});
httpsListener.addCertificates(`otlp-http-certificate`, [certificate]);

httpsListener.addTargets('otlp-http-target', {
  port: 4318,
  protocol: elb.ApplicationProtocol.HTTP,
  protocolVersion: elb.ApplicationProtocolVersion.HTTP1,
  targets: [autoScalingGroup],
  healthCheck: {
    protocol: elb.Protocol.HTTP,
    port: '13133',
    path: '/status',
    healthyHttpCodes: '200'
  }
});
```

```yaml
extensions:
  healthcheckv2:
    use_v2: true
    grpc:
      endpoint: "0.0.0.0:13132"
      transport: "tcp"
    http:
      endpoint: "0.0.0.0:13133"
      status:
        enabled: true
      config:
        enabled: false
```

mwear committed:
Post rebase updates

mwear committed:
Enable goleak

klyalldr commented:
Hi, I am looking forward to this improvement. 

Do you have an ETA when it will be released?  ... thanks

mwear commented:
@klyalldr, I will bring this up at the collector SIG tomorrow. I've been asking for reviews from approvers, but there is resistance due to the size of this PR. I have been asked to break it up. The first chunk is here: #32523.

gjshao44 commented:
@mwear, thanks for the tremendous work so far, looking forward to the new chunks getting sorted out

mwear committed:
Update go.mod

mwear committed:
Post rebase fixes

mwear commented:
I'm continuing to keep this reference PR up to date as it's being broken into smaller PRs for review. The previous chunk was: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32695, the current one is: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33528.

mwear commented:
I'm reopening this for purpose of reference, but do not expect anyone to review it.

mwear committed:
Check listen error

atoulme commented on .github/CODEOWNERS:
```diff
@@ -299,10 +299,10 @@ testbed/mockdatasenders/mockdatadogagentexporter/                   @open-teleme
 ## UNMAINTAINED components
 
-receiver/googlecloudspannerreceiver/                                @open-telemetry/collector-contrib-approvers 
+receiver/googlecloudspannerreceiver/                                @open-telemetry/collector-contrib-approvers
```
Please revert those white space changes, this file is generated

mwear replied:
Will do. FYI, this PR is just a reference. I've been keeping it up to date as I split off smaller PRs and they are integrated.

mwear commented on .github/CODEOWNERS:
```diff
@@ -299,10 +299,10 @@ testbed/mockdatasenders/mockdatadogagentexporter/                   @open-teleme
 #####################################################
-reports/distributions/core.yaml    @open-telemetry/collector-contrib-approvers 
-reports/distributions/contrib.yaml @open-telemetry/collector-contrib-approvers 
+reports/distributions/core.yaml    @open-telemetry/collector-contrib-approvers
```
```suggestion
reports/distributions/core.yaml    @open-telemetry/collector-contrib-approvers 
```

mwear commented on .github/CODEOWNERS:
```diff
@@ -299,10 +299,10 @@ testbed/mockdatasenders/mockdatadogagentexporter/                   @open-teleme
-reports/distributions/core.yaml    @open-telemetry/collector-contrib-approvers 
-reports/distributions/contrib.yaml @open-telemetry/collector-contrib-approvers 
+reports/distributions/core.yaml    @open-telemetry/collector-contrib-approvers
+reports/distributions/contrib.yaml @open-telemetry/collector-contrib-approvers
```
```suggestion
reports/distributions/contrib.yaml @open-telemetry/collector-contrib-approvers 
```

mwear commented on .github/CODEOWNERS:
```diff
@@ -299,10 +299,10 @@ testbed/mockdatasenders/mockdatadogagentexporter/                   @open-teleme
 ## UNMAINTAINED components
 
-receiver/googlecloudspannerreceiver/                                @open-telemetry/collector-contrib-approvers 
+receiver/googlecloudspannerreceiver/                                @open-telemetry/collector-contrib-approvers
```
```suggestion
receiver/googlecloudspannerreceiver/                                @open-telemetry/collector-contrib-approvers 
```

mwear committed:
whitespace

mwear commented:
This PR was broken down into smaller pieces, all of which have been reviewed and merged. healthcheckv2extension is available of anyone who wants to build their own collector, but not yet available in contrib. There is some refactoring of component status happening in core currently. I will aim to keep this extension working as that work is being done, but there could be some churn in the sort term. Currently the readme says the extension is not ready for use. However, it is ready for experimentation with the caveats I've just discussed. I will update the readme once the refactor in core is finished. I'm going to close this now that everything exists on main.

julealgon commented:
@mwear , I saw in the latest release notes that the `healthcheck` extension was in unmaintained state, and after following some links from the documentation, I landed on this PR here which was basically introducing a more comprehensive health checking mechanism.

You mentioned most of the changes have been merged, but that there is still some work to do. I see that v2 is still marked experimental to this day.

Is there an issue that tracks that remaining progress that one could subscribe to? We are currently using the old health checking extension and I would definitely like to switch to the new one as soon as it becomes available, but without a place to track its progress I'm a bit in the dark on this one.
